tensorpack/tfutils/argscope.py

```diff
@@ -64,26 +65,30 @@ def get_arg_scope():
         return defaultdict(dict)
 
 
-def argscope_mapper(func):
+def argscope_mapper(func, log_shape=True):
     """Decorator for function to support argscope
     """
     @wraps(func)
     def wrapped_func(*args, **kwargs):
         actual_args = copy.copy(get_arg_scope()[func.__name__])
         actual_args.update(kwargs)
         out_tensor = func(*args, **actual_args)
+        if log_shape:
+            in_tensor = args[0]
```

Comment:
I can add here something similar to https://github.com/tensorpack/tensorpack/blob/fa34d239ce0a6d5cc12bcb12ad690f07df5f1f0e/tensorpack/models/registry.py#L118-L119

But this would introduce a tensorflow as a dependency to this file.

Reply:
Yeah it's best to make multigpu training only print shapes in the first tower, but I'm OK without this feature. It's fine for files in `tfutils` to depend on tf.

In fact I'm not sure how variable/name_scopes in tf.layers work.. I remember it's quite complicated..

Reply:
Indeed, this is not trivial as I cannot access the scope inside `tf.layers.lowercase_name_function`.

The lastest commit in this PR at least can distinguish between the towers

```python
from tensorpack import *
import tensorflow as tf

enable_argscope_for_module(tf.layers)

def build(x):
    return tf.layers.conv2d(x, 32, [3, 3], name='conv0')
ans = []

print "----------------------------------"
plhdr = tf.placeholder(tf.float32, [1, 28, 28, 3])
ans.append(build(plhdr))

print "----------------------------------"
with TowerContext('tower0', True, vs_name='tower0') as ctx:
    ans.append(build(plhdr))

print "----------------------------------"
with TowerContext('tower1', True, vs_name='tower1') as ctx:
    ans.append(build(plhdr))

print "----------------------------------"
with TowerContext('tower2', True, vs_name='tower2') as ctx:
    ans.append(build(plhdr))


```

Reply:
The `InferenceTower` is still dumped to the console. Which is ok, as it might differ from the training tower.